**NETWORK INTRUSION DETECTION SYSTEM**

In [1]:
#Importing Libraries
import pandas as pd
pd.set_option('display.width', 10000)  # Width of the display in characters
pd.set_option('display.max_rows', 100)  # Maximum number of rows to display
pd.set_option('display.max_columns', 100)  # Maximum number of columns to display
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
from sklearn.metrics import f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix, classification_report
from sklearn import metrics # Importing 'metrics' from sklearn
from imblearn.over_sampling import SMOTE
from keras.utils import to_categorical
from sklearn.ensemble import BaggingClassifier
import warnings
# Ignore all warnings
warnings.filterwarnings('ignore')

*   The code mounts Google Drive to access files and directories from within Google Colab.




In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
# Reads a CSV file containing training data from Google Drive
train = pd.read_csv('/content/drive/MyDrive/UNSW_NB15_training-set.csv')
#Displays the first few rows of the DataFrame.
train.head()

id       dur proto service state  spkts  dpkts  sbytes  dbytes       rate  sttl  dttl         sload          dload  sloss  dloss      sinpkt      dinpkt          sjit          djit  swin       stcpb       dtcpb  dwin    tcprtt    synack    ackdat  smean  dmean  trans_depth  response_body_len  ct_srv_src  ct_state_ttl  ct_dst_ltm  ct_src_dport_ltm  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  ct_ftp_cmd  ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports attack_cat  label
0   1  0.121478   tcp       -   FIN      6      4     258     172  74.087490   252   254  14158.942380    8495.365234      0      0   24.295600    8.375000     30.177547     11.830604   255   621772692  2202533631   255  0.000000  0.000000  0.000000     43     43            0                  0           1             0           1                 1                 1               1             0           0                 0           1           1                0     Normal      0
1   2  0.649902   tcp       -   FIN     14     38     734   42014  78.473372    62   252   8395.112305  503571.312500      2     17   49.915000   15.432865     61.426934   1387.778330   255  1417884146  3077387971   255  0.000000  0.000000  0.000000     52   1106            0                  0          43             1           1                 1                 1               2             0           0                 0           1           6                0     Normal      0
2   3  1.623129   tcp       -   FIN      8     16     364   13186  14.170161    62   252   1572.271851   60929.230470      1      6  231.875571  102.737203  17179.586860  11420.926230   255  2116150707  2963114973   255  0.111897  0.061458  0.050439     46    824            0                  0           7             1           2                 1                 1               3             0           0                 0           2           6                0     Normal      0
3   4  1.681642   tcp     ftp   FIN     12     12     628     770  13.677108    62   252   2740.178955    3358.622070      1      3  152.876547   90.235726    259.080172   4991.784669   255  1107119177  1047442890   255  0.000000  0.000000  0.000000     52     64            0                  0           1             1           2                 1                 1               3             1           1                 0           2           1                0     Normal      0
4   5  0.449454   tcp       -   FIN     10      6     534     268  33.373826   254   252   8561.499023    3987.059814      2      1   47.750333   75.659602   2415.837634    115.807000   255  2436137549  1977154190   255  0.128381  0.071147  0.057234     53     45            0                  0          43             1           2                 2                 1              40             0           0                 0           2          39                0     Normal      0

In [5]:
# Replacing '-' with NaN in column 'Service'
train['service'] = train['service'].replace('-', pd.NA)

# Calculating the mode value of column 'Service' (ignoring NaN values)
mode_value = train['service'].mode(dropna=True)[0]

# Replacing NaN with the mode value in column 'Service'
train['service'] = train['service'].fillna(mode_value)

In [6]:
# Drops the 'id' and 'label' columns from the DataFrame 'train'
train.drop(columns=["id","label"], inplace=True)

In [7]:
# Displaying the first 5 rows after replacing nans with mode value in service column
train.head()

dur proto service state  spkts  dpkts  sbytes  dbytes       rate  sttl  dttl         sload          dload  sloss  dloss      sinpkt      dinpkt          sjit          djit  swin       stcpb       dtcpb  dwin    tcprtt    synack    ackdat  smean  dmean  trans_depth  response_body_len  ct_srv_src  ct_state_ttl  ct_dst_ltm  ct_src_dport_ltm  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  ct_ftp_cmd  ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports attack_cat
0  0.121478   tcp     dns   FIN      6      4     258     172  74.087490   252   254  14158.942380    8495.365234      0      0   24.295600    8.375000     30.177547     11.830604   255   621772692  2202533631   255  0.000000  0.000000  0.000000     43     43            0                  0           1             0           1                 1                 1               1             0           0                 0           1           1                0     Normal
1  0.649902   tcp     dns   FIN     14     38     734   42014  78.473372    62   252   8395.112305  503571.312500      2     17   49.915000   15.432865     61.426934   1387.778330   255  1417884146  3077387971   255  0.000000  0.000000  0.000000     52   1106            0                  0          43             1           1                 1                 1               2             0           0                 0           1           6                0     Normal
2  1.623129   tcp     dns   FIN      8     16     364   13186  14.170161    62   252   1572.271851   60929.230470      1      6  231.875571  102.737203  17179.586860  11420.926230   255  2116150707  2963114973   255  0.111897  0.061458  0.050439     46    824            0                  0           7             1           2                 1                 1               3             0           0                 0           2           6                0     Normal
3  1.681642   tcp     ftp   FIN     12     12     628     770  13.677108    62   252   2740.178955    3358.622070      1      3  152.876547   90.235726    259.080172   4991.784669   255  1107119177  1047442890   255  0.000000  0.000000  0.000000     52     64            0                  0           1             1           2                 1                 1               3             1           1                 0           2           1                0     Normal
4  0.449454   tcp     dns   FIN     10      6     534     268  33.373826   254   252   8561.499023    3987.059814      2      1   47.750333   75.659602   2415.837634    115.807000   255  2436137549  1977154190   255  0.128381  0.071147  0.057234     53     45            0                  0          43             1           2                 2                 1              40             0           0                 0           2          39                0     Normal

In [8]:
# Reads a CSV file containing training data from Google Drive
test = pd.read_csv('/content/drive/MyDrive/UNSW_NB15_testing-set.csv')
# Displays the first few rows of the DataFrame
test.head()

id       dur proto service state  spkts  dpkts  sbytes  dbytes         rate  sttl  dttl        sload  dload  sloss  dloss  sinpkt  dinpkt  sjit  djit  swin  stcpb  dtcpb  dwin  tcprtt  synack  ackdat  smean  dmean  trans_depth  response_body_len  ct_srv_src  ct_state_ttl  ct_dst_ltm  ct_src_dport_ltm  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  ct_ftp_cmd  ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports attack_cat  label
0   1  0.000011   udp       -   INT      2      0     496       0   90909.0902   254     0  180363632.0    0.0      0      0   0.011     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    248      0            0                  0           2             2           1                 1                 1               2             0           0                 0           1           2                0     Normal      0
1   2  0.000008   udp       -   INT      2      0    1762       0  125000.0003   254     0  881000000.0    0.0      0      0   0.008     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    881      0            0                  0           2             2           1                 1                 1               2             0           0                 0           1           2                0     Normal      0
2   3  0.000005   udp       -   INT      2      0    1068       0  200000.0051   254     0  854400000.0    0.0      0      0   0.005     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    534      0            0                  0           3             2           1                 1                 1               3             0           0                 0           1           3                0     Normal      0
3   4  0.000006   udp       -   INT      2      0     900       0  166666.6608   254     0  600000000.0    0.0      0      0   0.006     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    450      0            0                  0           3             2           2                 2                 1               3             0           0                 0           2           3                0     Normal      0
4   5  0.000010   udp       -   INT      2      0    2126       0  100000.0025   254     0  850400000.0    0.0      0      0   0.010     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0   1063      0            0                  0           3             2           2                 2                 1               3             0           0                 0           2           3                0     Normal      0

In [9]:
# Replacing '-' with NaN in column 'Service'
test['service'] = test['service'].replace('-', pd.NA)

# Calculating the mode value of column 'Service' (ignoring NaN values)
mode_value_test = test['service'].mode(dropna=True)[0]

# Replacing NaN with the mode value in column 'Service'
test['service'] = test['service'].fillna(mode_value_test)

In [10]:
# Drops the 'id' and 'label' columns from the DataFrame 'test'
test.drop(columns=["id","label"], inplace=True)

In [11]:
# Displaying the first 5 rows after replacing nans with mode value in service column
test.head()

dur proto service state  spkts  dpkts  sbytes  dbytes         rate  sttl  dttl        sload  dload  sloss  dloss  sinpkt  dinpkt  sjit  djit  swin  stcpb  dtcpb  dwin  tcprtt  synack  ackdat  smean  dmean  trans_depth  response_body_len  ct_srv_src  ct_state_ttl  ct_dst_ltm  ct_src_dport_ltm  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  ct_ftp_cmd  ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports attack_cat
0  0.000011   udp     dns   INT      2      0     496       0   90909.0902   254     0  180363632.0    0.0      0      0   0.011     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    248      0            0                  0           2             2           1                 1                 1               2             0           0                 0           1           2                0     Normal
1  0.000008   udp     dns   INT      2      0    1762       0  125000.0003   254     0  881000000.0    0.0      0      0   0.008     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    881      0            0                  0           2             2           1                 1                 1               2             0           0                 0           1           2                0     Normal
2  0.000005   udp     dns   INT      2      0    1068       0  200000.0051   254     0  854400000.0    0.0      0      0   0.005     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    534      0            0                  0           3             2           1                 1                 1               3             0           0                 0           1           3                0     Normal
3  0.000006   udp     dns   INT      2      0     900       0  166666.6608   254     0  600000000.0    0.0      0      0   0.006     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0    450      0            0                  0           3             2           2                 2                 1               3             0           0                 0           2           3                0     Normal
4  0.000010   udp     dns   INT      2      0    2126       0  100000.0025   254     0  850400000.0    0.0      0      0   0.010     0.0   0.0   0.0     0      0      0     0     0.0     0.0     0.0   1063      0            0                  0           3             2           2                 2                 1               3             0           0                 0           2           3                0     Normal

In [12]:
# Checking if there are any null values in each column in train data.
train.isnull().any()

dur                  False
proto                False
service              False
state                False
spkts                False
dpkts                False
sbytes               False
dbytes               False
rate                 False
sttl                 False
dttl                 False
sload                False
dload                False
sloss                False
dloss                False
sinpkt               False
dinpkt               False
sjit                 False
djit                 False
swin                 False
stcpb                False
dtcpb                False
dwin                 False
tcprtt               False
synack               False
ackdat               False
smean                False
dmean                False
trans_depth          False
response_body_len    False
ct_srv_src           False
ct_state_ttl         False
ct_dst_ltm           False
ct_src_dport_ltm     False
ct_dst_sport_ltm     False
ct_dst_src_ltm       False
is_ftp_login         False
c

No missing values in any column of the train data

In [13]:
# Checking if there are any null values in each column in test data.
test.isnull().any()

dur                  False
proto                False
service              False
state                False
spkts                False
dpkts                False
sbytes               False
dbytes               False
rate                 False
sttl                 False
dttl                 False
sload                False
dload                False
sloss                False
dloss                False
sinpkt               False
dinpkt               False
sjit                 False
djit                 False
swin                 False
stcpb                False
dtcpb                False
dwin                 False
tcprtt               False
synack               False
ackdat               False
smean                False
dmean                False
trans_depth          False
response_body_len    False
ct_srv_src           False
ct_state_ttl         False
ct_dst_ltm           False
ct_src_dport_ltm     False
ct_dst_sport_ltm     False
ct_dst_src_ltm       False
is_ftp_login         False
c

No missing values in any column of the test data

In [14]:
#Checking Shape of train data
train.shape

(175341, 43)

* There are 175,341 records and 43 columns in train data




In [15]:
#Checking shape of test data
test.shape

(82332, 43)



* There are 82332 records and 43 columns in test data




In [ ]:
#calculates the counts of unique values in the 'attack_cat' column of the DataFrame 'train'
category_counts = train["attack_cat"].value_counts()
# Print the count of each category
print(category_counts)

attack_cat
Normal            56000
Generic           40000
Exploits          33393
Fuzzers           18184
DoS               12264
Reconnaissance    10491
Analysis           2000
Backdoor           1746
Shellcode          1133
Worms               130
Name: count, dtype: int64






*  For both train and test data, the below mentioned attack categories are less in number comparitively.

1.   Analysis
2.   Backdoor
3.   Shellcode
4.   Worms

*   Hence deleting all these categories for both train and test data.


In [16]:
l = ['Analysis','Backdoor','Shellcode','Worms']
for i in l:
    train = train[train.attack_cat.isin(l) == False]
    test = test[test.attack_cat.isin(l) == False]

In [17]:
category_counts = train["attack_cat"].value_counts()
# Print the count of each category
print(category_counts)

attack_cat
Normal            56000
Generic           40000
Exploits          33393
Fuzzers           18184
DoS               12264
Reconnaissance    10491
Name: count, dtype: int64


In [18]:
# Here attack_cat column is the target variable
y_train = train['attack_cat']
y_test  = test['attack_cat']